### ISSUE TITLE
[Feature]: iOS Core Bluetooth Background Processing support

### ISSUE THREAD

**Issue by William-Geng** [boskokg/flutter_blue_plus#846]
### FlutterBluePlus Version

1.31.17

### Flutter Version

3.19.3

### What OS?

iOS

### OS Version

17

### Bluetooth Module

nordic nRF52832

### What is your feature request?

It would be nice to support iOS core bluetooth background execution mode as described here: https://developer.apple.com/library/archive/documentation/NetworkingInternetWeb/Conceptual/CoreBluetooth_concepts/CoreBluetoothBackgroundProcessingForIOSApps/PerformingTasksWhileYourAppIsInTheBackground.html.

I believe there is a PR for flutter_blue that was never merged:  https://github.com/pauldemarco/flutter_blue/pull/191 and https://github.com/pauldemarco/flutter_blue/pull/210.

### Logs

```shell
none
```

**Comment by chipweinberger** [comment]
feel free to create new PRs and test them

**Comment by MrCsabaToth** [comment]
FYI there was a https://github.com/espresso3389/flutter_blue fork of the original pauldemarco `flutter_blue` back in the day which dealt with foreground task issue as well, along with some emphasis on iOS and also permissions. But seems like that fork died too some times after `flutter_blue`, but it might be useful to see what PRs they merged and what they did for this.

I'll need some foreground support as well in my app at some point - highly requested -, however this seems to be a major undertaking in my case. More specifically at my app level this looks to be a significant architectural change. Even though there are plugins such as https://pub.dev/packages/flutter_foreground_task/ you can see it's unfortunately not a simple plop in of some code, it's 5+ stage step procedure. I also don't see a way once I convert my app, to have that functionality be optional. That's important because if such function would be at the FBP plugin level, it could happen not everyone would want it (because for example it needs extra manifest declarations, you'd need to select type of service, extra permission shenanigans, declaring callbacks, and the list goes on and on).

But ultimately I'd need to tackle this as well for user request. I'm not sure if it's even doable at FBP level, it'll be hard enough at app level. I'm curious what is your use-case and plans.

**Comment by Codel1417** [comment]
My initial testing with the app in the background, using a 15 second keep awake ping, kept my app open.

**Comment by h4nkyn** [comment]
What type of keep awake ping are you talking about? Like a characteristic value change?

**Comment by Codel1417** [comment]
> What type of keep awake ping are you talking about? Like a characteristic value change?

Yeah. Having a characteristic value change wakes the app for 10-15 seconds. If you do this enough the app stays open

**Comment by drolpi** [comment]
> feel free to create new PRs and test them

@chipweinberger are there any updates on this? I absolutely need this feature! Otherwise, I have to completely rebuild my app in a different language with a different framework. Can someone please implement IOS BLE State Restoration.

**Comment by chipweinberger** [comment]
@drolpi , just open a PR!

**Comment by chipweinberger** [comment]
@drolpi 

> Can someone please implement IOS BLE State Restoration.

I don't need this feature, but I will implement it for a fee. email me.

 Otherwise you should implement it yourself.

**Comment by MrCsabaToth** [comment]
> @chipweinberger are there any updates on this? I absolutely need this feature! Otherwise, I have to completely rebuild my app in a different language with a different framework. Can someone please implement IOS BLE State Restoration.

BLE State Restoration and background services are two different things, although may not be completely unrelated (in case the the background service gets yeeted (this part of Android is not standardized https://dontkillmyapp.com/) then restoration may come into play?)

**Comment by MrCsabaToth** [comment]
I think Chip did some work related to reconnection if I'm not mistaken, however BLE State Restoration sounds more formal.

**Comment by chipweinberger** [comment]
I've pushed this commit (https://github.com/chipweinberger/flutter_blue_plus/commit/12c17a6f49c3d6223f7dffaa75f845ccc01c8701) , which *should* enable this feature. But I have not been able to get it to work in actual testing.

1. when the app is killed, it is not restored
2. when the phone is rebooted, it is not restored

**Comment by chipweinberger** [comment]
released version 1.32.13 with State Preservation

**Comment by Scottie-Fischer** [comment]
> * when the app is killed, the app is not restored ❌
> * when the phone is rebooted, the app is not restored ❌

I made some minor tweaks using your latest update and was able to get it to restore the CBCentralManager and peripherals. 
I did see some issues where when I restore the peripheral and save it in the `connectedDevices` list that it times out when trying to discover services or subscribe to characteristic using `setNotifyingValue`. Have not pinned down the issues, but will make a PR. 

Additionally I created a restoration state that tells the app whether or not it was restored or the initial start. This may be useful for diverging behaviors, but not technically required.

**Comment by chipweinberger** [comment]
awesome!

i look forward to the PR!

**Comment by gpminsuk** [comment]
@chipweinberger Thanks for this update. I was able to verify that the app reconnect, wake up, and runs the code.

> when the app is killed, the app is not restored ❌

When you say this, do you mean when the app is killed by user? Or do you mean when the app is OS as well?
The problem I am having now is that
1. My device is connected with state restoration and autoconnect
2. The device gets disconnected due to distance from the central device (app).
3. Time passes and OS decides to kill the app.
4. The device come back close to the central device (app)
5. Reconnect / wake up does not happen.

I believe iOS should support this scenario. If you could shed some light, I can try to prepare a PR.

**Comment by RohitMishra1402** [comment]
Is there any working solution for background fetching BLE for iOS devices?

**Comment by chipweinberger** [comment]
not sure. can someone else chime in?

**Comment by RohitMishra1402** [comment]
@all Is there any documentation stating that background BLE fetching on iOS devices is impossible?

**Comment by tayz98** [comment]
@Codel1417 

> > What type of keep awake ping are you talking about? Like a characteristic value change?
> 
> Yeah. Having a characteristic value change wakes the app for 10-15 seconds. If you do this enough the app stays open. The device I connect to uses 2 characteristics for a read/write command setup so I just sent ping which gets a response.


Sorry for digging up this thread. Basically this is what i am doing but I have the feeling that even when the app is alive it doesn't execute any code that is not related to bluetooth. I need to create objects and store the collected BLE data, but the app is just skipping/ignoring these commands only executing "write commands", which is basically worthless if i can't do anything with the received data....

**Comment by gpminsuk** [comment]
I am looking for someone who can help resolve this issue as a paid task.

**Comment by iyzg** [comment]
Me as well. Can background BLE write to characteristics or scan for devices?

**Comment by cyonix** [comment]
> I made some minor tweaks using your latest update and was able to get it to restore the CBCentralManager and peripherals.
> I did see some issues where when I restore the peripheral and save it in the `connectedDevices` list that it times out when trying to discover services or subscribe to characteristic using `setNotifyingValue`. Have not pinned down the issues, but will make a PR.

@Scottie-Fischer Did you end up creating the PR? I'm interested in seeing what changes you made here.

**Comment by shimon88** [comment]
Hi All,
According to the documentation about working in the background in iOS "This is an advanced use case. FlutterBluePlus does not support everything. You may have to fork it. PRs are welcome.".
we are in the process of evaluate the feasibility of our solution which requires connecting and downloading data from a device all while the app is in the background without user interaction.
We would very appreciate any details on what is not implemented yet in the package? what should we expect when working with this package and what are your though on using it outside of development enviroments?

any assistant and guidance will be greatly appreciated

**Comment by chipweinberger** [comment]
This ChatGPT convo was useful: https://chatgpt.com/share/68a6ef12-7de8-8011-8245-3c97cdc1e5bc

-------

## **To ChatGPT**:  "as an app developer, what am i allowed to do with core bluetooth in the background?"

-------

On iOS, **Core Bluetooth** has strict rules for what you can and cannot do in the background. Apple only allows background Bluetooth use in specific scenarios, and you must declare the appropriate background modes in your app’s `Info.plist`.

Here’s the breakdown:

## ✅ What You’re Allowed to Do

If you enable the **"Uses Bluetooth LE accessories"** background mode (`bluetooth-central`, `bluetooth-peripheral`):

### As a Central (scanning/connecting to peripherals)

* **Continue scanning for peripherals** in the background (with restrictions: results may be delayed and not as frequent as in the foreground).
* **Maintain active connections** to peripherals while in the background.
* **Receive and process updates** (e.g. notifications/indications on characteristics) from connected peripherals.
* **Reconnect automatically** to previously connected peripherals when they advertise again (if you use `CBCentralManager` with state restoration).

### As a Peripheral (advertising to centrals)

* **Continue advertising** while in the background, but with limits:

  * Advertising data is reduced (only a service UUID is sent, no custom data).
  * Advertising frequency is reduced to save power.
* **Respond to read/write requests** from connected centrals.
* **Send notifications/indications** to subscribed centrals.

## 🚫 What You’re *Not* Allowed to Do

* **Unlimited background scanning**: You cannot continuously discover all nearby BLE devices. Apple throttles scan results and only delivers them under certain conditions.
* **Custom advertising payloads in background**: You cannot advertise arbitrary manufacturer data or long payloads — only a limited service UUID list.
* **Run arbitrary tasks** in the background: Core Bluetooth background execution is limited to Bluetooth-related work; you can’t keep your app alive to run unrelated code.
* **Keep the CPU awake indefinitely**: Your app wakes only for Bluetooth-related events, not for general processing.

## ⚠️ Practical Considerations

* You must enable background modes in **Xcode → Signing & Capabilities → Background Modes**.
* For **state restoration**, you should provide identifiers when initializing `CBCentralManager` / `CBPeripheralManager`. iOS will relaunch your app in the background if a relevant Bluetooth event occurs.
* If you abuse background Bluetooth (e.g. continuous scanning to approximate location), Apple may **reject your app** during App Review.

✅ **In short:** You can keep connections alive, exchange data, advertise with restrictions, and wake up for Bluetooth events. ❌ You cannot use Bluetooth as a loophole to run your app continuously in the background or broadcast arbitrary data.

---
## **To ChatGPT**: "tell me how Receiving and processing updates works"
---


Great question — this is one of the **most useful Core Bluetooth background behaviors**, and it’s also one of the least well-documented. Let me walk you through how it actually works under the hood.

## 1. What “updates” means

In BLE, a peripheral (your accessory) can **notify** or **indicate** changes in the value of a characteristic.

* **Notification** = “fire and forget” → peripheral pushes new data, central (your iOS app) receives it, no acknowledgment required.
* **Indication** = like notification but requires the central to acknowledge receipt.

When your iOS app subscribes to these updates (via `setNotifyValue(true, for: characteristic)`), the peripheral pushes data whenever the value changes.

## 2. What happens in the background

If your app:

* Has **“Uses Bluetooth LE accessories”** background mode enabled, and
* Is **connected** to a peripheral,

then iOS will **keep the BLE connection alive in the background**.

* The peripheral sends a notification/indication.
* iOS wakes your app (even if it’s suspended) just long enough to deliver `peripheral:didUpdateValueForCharacteristic:error:`.
* Your app can process the data, save it (e.g. to Core Data, UserDefaults, or a file), then it’s suspended again until the next event.

## 3. Lifecycle + State Restoration

* If your app was **force quit by the system** (not swiped away by the user), iOS can relaunch it in the background when the peripheral sends data.
* This works if you initialized `CBCentralManager` with a **restoration identifier**. On relaunch, iOS gives you the list of peripherals and state so you can resume.

Example:

```swift
centralManager = CBCentralManager(delegate: self,
                                  queue: nil,
                                  options: [CBCentralManagerOptionRestoreIdentifierKey: "MyCentralManager"])
```

Then implement:

```swift
func centralManager(_ central: CBCentralManager,
                    willRestoreState dict: [String : Any]) {
    // dict contains the peripherals you were connected to
}
```

## 4. Power and delivery constraints

* Notifications **do not wake the screen**; they just wake the app in the background for processing.
* Delivery can be **slower** in background (batched) if the system decides to save power, but it’s usually quite prompt for active connections.
* If the user turns off Bluetooth, or your app loses connection, no more updates come until you reconnect.

## 5. Typical use cases

* Health sensors (heart rate, glucose monitor, etc.) continuously streaming data.
* MIDI/BLE musical devices (like your Jamcorder 🎹) sending note events.
* IoT sensors periodically updating readings.

👉 In practice: if your peripheral keeps pushing characteristic updates, your app will **wake, receive them, and process them in the background**, even when the user hasn’t opened your app in hours.

---
## **To ChatGPT**:  **how does this work in the context of Flutter? will my whole flutter application relaunch? how can i receive the data and process it"
---

Short answer: on iOS your **app process can be relaunched in the background**, but your **Flutter (Dart) layer is not automatically running** unless you arrange for it. CoreBluetooth events are delivered first to **native iOS code** (Swift/Obj-C). The reliable pattern is:

1. let native iOS receive notifications/indications while the app is backgrounded or relaunched,
2. persist/buffer that data natively, and
3. hand it to Flutter/Dart when the UI/isolate is alive (or spin up a small, headless Flutter engine yourself to do quick processing).

Below is a concrete, battle-tested way to wire this up.

# What iOS does for you

* If you declare `bluetooth-central` in **Background Modes** and give your `CBCentralManager` a **restoration identifier**, iOS will:

  * keep active BLE connections alive in background,
  * wake your process to deliver `didUpdateValueForCharacteristic`,
  * and even **relaunch** your app in the background after a crash/OS kill (not after a user swipe-quit).
* Those callbacks happen in **native code**. Your Flutter isolate is usually suspended.

# Architecture that works with Flutter

## A. Native receives → buffer → later deliver to Flutter

**Best for reliability and App Review friendliness.**

1. **Enable background mode & restoration**

   * Xcode > Signing & Capabilities > Background Modes → check **Uses Bluetooth LE accessories**.
   * Create your central with a restoration id:

   ```swift
   central = CBCentralManager(
     delegate: self,
     queue: nil,
     options: [CBCentralManagerOptionRestoreIdentifierKey: "BLECentralRestoreID"]
   )
   ```

2. **Handle restoration and updates**

   ```swift
   // AppDelegate or your plugin Swift class
   func centralManager(_ central: CBCentralManager, willRestoreState dict: [String : Any]) {
       // Reacquire peripherals and set delegates
   }

   func peripheral(_ peripheral: CBPeripheral,
                   didUpdateValueFor characteristic: CBCharacteristic,
                   error: Error?) {
       guard error == nil, let data = characteristic.value else { return }
       // ❶ Persist quickly: append to a file/UserDefaults/SQLite/CoreData
       BLEBuffer.shared.append(peripheral: peripheral.identifier, char: characteristic.uuid, payload: data)
       // ❷ If Flutter is currently running, also push via EventChannel
       FlutterBridge.shared.emitIfReady(event: dataPayload)
   }
   ```

   Implementation notes:

   * Keep work **very short**. If you need a few extra seconds, wrap work in `beginBackgroundTask(expirationHandler:)`.
   * Prefer a **file / SQLite** ring buffer over giant `UserDefaults` blobs.

3. **Flush to Flutter when the UI is alive**

   * On app foreground / Flutter engine ready, read buffered items and send them to Dart via a **MethodChannel/EventChannel**.

   ```swift
   let channel = FlutterMethodChannel(name: "ble_stream", binaryMessenger: messenger)
   // on Dart “ready”, native drains BLEBuffer and sends the backlog
   ```

4. **Dart side**

   ```dart
   // Event stream for live updates when app is foregrounded
   final events = const EventChannel('ble_stream_events').receiveBroadcastStream();

   // MethodChannel to request a backlog flush on startup/resume
   final method = const MethodChannel('ble_stream');
   Future<void> pullBacklogOnResume() async {
     final List<dynamic> items = await method.invokeMethod('drainBacklog');
     for (final item in items) processBlePacket(item);
   }
   ```

Pros: rock-solid delivery; works even when Flutter isn’t running.
Cons: you’re doing some native code (but it’s thin).

## B. Start a **headless Flutter engine** for quick background processing

**Optional**, if you truly need Dart to run in the background on iOS.

* On any BLE wake (including restoration), you can spin up a **secondary FlutterEngine** (no UI), run a specific Dart entrypoint (e.g., `bleBackgroundMain`), pass the payload over a `MethodChannel`, do your work quickly, then tear down the engine.

Swift:

```swift
func runHeadlessDart(with payload: Data) {
    let engine = FlutterEngine(name: "ble_bg_engine", project: nil, allowHeadlessExecution: true)
    engine.run(withEntrypoint: "bleBackgroundMain")
    let channel = FlutterMethodChannel(name: "ble_bg", binaryMessenger: engine.binaryMessenger)
    channel.invokeMethod("onBlePayload", arguments: payload.base64EncodedString())
    // End the engine when done; keep within iOS background time limits
}
```

Dart:

```dart
@pragma('vm:entry-point')
void bleBackgroundMain() async {
  const ch = MethodChannel('ble_bg');
  ch.setMethodCallHandler((call) async {
    if (call.method == 'onBlePayload') {
      final data = base64Decode(call.arguments as String);
      await processAndPersist(data); // keep it quick!
    }
  });
}
```

Caveats:

* You **must** keep work short; iOS grants limited background time.
* Not all plugins are safe in a headless engine.
* This is more complex than buffering natively.

# Notes about popular Flutter BLE plugins

* **flutter\_blue\_plus** / **flutter\_reactive\_ble**: on iOS they rely on CoreBluetooth under the hood. They work great when your app is foregrounded. For **background delivery while suspended**, you still need the native buffering or a headless engine approach above. Don’t expect a pure-Dart listener to fire while the app is fully suspended.

# Gotchas & limits

* **User swipe-quit**: iOS won’t relaunch your app until the user opens it again.
* **Don’t do heavy work** in `didUpdateValueForCharacteristic`—persist and exit.
* **Testing**: use a real device; background Bluetooth behavior is different in the simulator.
* **App Review**: Only use BLE background for **genuine accessory communication**; avoid using it as a keep-alive.

## Minimal checklist

* [ ] Add `bluetooth-central` to Background Modes.
* [ ] Create `CBCentralManager` with a restoration ID.
* [ ] In native, implement `willRestoreState` and `didUpdateValueForCharacteristic`.
* [ ] Persist updates immediately to a local buffer.
* [ ] On Flutter resume/launch, drain the buffer via MethodChannel/EventChannel.
* [ ] (Optional) Headless Flutter engine for short background Dart work.

**Comment by chipweinberger** [comment]
It seems like the piece needed to get things running in Dart is this: **B. Start a headless Flutter engine for quick background processing**

**Comment by gpminsuk** [comment]
Did anyone successfully using this project for BLE connection restoration and app wake up?

Here's my scenario.

1. My app scans the device using Service UUID
2. Device is detected and establish connection
3. User swipe the App to background
4. For 10+ hours, BLE connection is maintained with no issue
5. Device gets out of range, therefore is disconnected.
6. Device starts advertising (but it's out of range so the app doesn't know)
7. After 10+ hours, device gets within the range. (at this time, I assume that the app is suspended or killed)

**Expected Behavior:** App wakes up and establish connection with the device
**Actual Behavior:** App does not wake up and the device keeps advertising

**Comment by chipweinberger** [comment]
Yes that should be possible if according to ChatGPT. Here is the critical line:

> If your app was force quit by the _system_ (not swiped away by the user), iOS can relaunch it in the background when the peripheral sends data.

**Comment by gpminsuk** [comment]
Yes, but what would be missing such that this library today doesn't do the expected behavior?

The scenario I described is the actual issue I am facing right now.

Is it this you mentioned above?
**B. Start a headless Flutter engine for quick background processing**

**Comment by chipweinberger** [comment]
> Yes, but what would be missing such that this library today doesn't do the expected behavior?

see here: https://github.com/chipweinberger/flutter_blue_plus/issues/846#issuecomment-3209917451

and here: https://github.com/chipweinberger/flutter_blue_plus/issues/846#issuecomment-3209882823

**Comment by cyonix** [comment]
@chipweinberger I think this commit: a2c5f05f4fc6b5a0218d9480360a4340333e9805 solves the CoreBluetooth restoration issue for darwin platforms. 

With this change, this use case works:
- when the phone is rebooted the app is automatically restored while in the background

Before this commit, the client would call `FlutterBluePlus.setOptions(restoreState: true);` which in turn caused `FlutterBluePlusPlugin` to initialize the `CBCentralManager` with an empty `options` dictionary. This means the central manager was missing the restoration identifier key required for restoration by iOS.

I'm not seeing this behavior after the commit.

**Note**: Restoration only occurs if the BLE device sends an event _after_ the phone reboots successfully.

**Comment by chipweinberger** [comment]
that commit was just fixing a regression since 1.35.0

this github issue has been open for much longer than that.

so no, i dont think that commit is all that is needed for a full implementation.

**Comment by shimon88** [comment]
Hi All,

i opened another issue but was instructed to talk about it here, please see #1291 
basically state restoration seems to work, the app wakes from notifications or connection, but connected devices list is empty
anyone has an idea?
